Feat/refactor user rates

valeryV99 opened the pull request:
https://www.notion.so/254362757c98816eaa18db14ba3bcb0d?v=254362757c9881b38afd000c1ae9b764&p=26a362757c988020be2ccffed6192858&pm=s

valeryV99 committed:
refactor(SwapForm): optimize rate fetching and clean up logic

- Replaced legacy rate fetching methods with `swapParams` for centralized type and value tracking.
- Updated `useSwapRates` to directly handle token pair parameters and handle missing args gracefully.
- Removed redundant disabled state for input fields during rate fetch.
- Simplified tests for `useSwapRates` by utilizing `useReadContract` mock.

valeryV99 committed:
Merge remote-tracking branch 'origin/staging' into feat/refactor-user-rates

valeryV99 committed:
refactor(SwapForm): simplify swap rates logic and improve state handling

- Replaced `useRef` with `useState` for `swapParams` to ensure reactivity.
- Removed unnecessary `useEffect` for rate updates; logic moved into debounced callbacks.
- Integrated `flushSync` for immediate updates of `swapParams`.
- Updated tests to align with refactored `useSwapRates` logic and improved mock argument verification.

valeryV99 committed:
Merge remote-tracking branch 'origin/staging' into feat/refactor-user-rates

# Conflicts:
#	apps/web/src/widgets/swap-form/ui/SwapForm.tsx

Georgyskr commented on apps/web/src/features/swap/api/useSwapRates.ts:
```diff
@@ -1,56 +1,45 @@
-    if (reverse) {
-      method = 'getAmountsIn';
-    }
+  const args = useMemo(() => {
```
just to double check, what's the point of useMemo here?

valeryV99 replied:
done

Georgyskr commented on apps/web/src/features/swap/api/useSwapRates.ts:
```diff
@@ -1,56 +1,45 @@
-      setIsFetching(false);
-    }
-  };
+  const { data, error, isFetching, refetch } = useReadContract({
```
I think return userReadContract would be sufficient

valeryV99 replied:
done

Georgyskr commented on apps/web/src/features/swap/api/useSwapRates.test.tsx:
```diff
@@ -24,79 +18,64 @@ vi.mock('@/global', () => ({
-  it('calls getAmountsIn when reverse=true', async () => {
-    readContractMock.mockResolvedValueOnce([456n, 0n]);
+    // Ensure hook passed correct params
+    const call = useReadContractMock.mock.calls.at(-1)?.[0];
```
let's use 'lastCall' as the code is hard to read atm.

or 
```
expect(mockFn).toHaveBeenCalledWith(...)
```

valeryV99 replied:
done

Georgyskr commented on apps/web/src/features/swap/api/useSwapRates.test.tsx:
```diff
@@ -24,79 +18,64 @@ vi.mock('@/global', () => ({
-      expect(result.current.isFetching).toBe(false);
-      expect(!!result.current.error).toBe(true);
-    });
+    renderHook(() => useSwapRates({ tokenIn: undefined, tokenOut: '0xB' as any, type: 'exactIn', value: undefined }));
```
Let's cover all missing args if undefined

Georgyskr commented on apps/web/src/widgets/swap-form/ui/SwapForm.tsx:
```diff
@@ -77,42 +78,43 @@ export const SwapForm = ({
-      formatUnits(outputAmount, outputTokenInfo.decimals),
-    );
+    const { data } = await readSwapRates();
+    const quote = data ? data[data.length - 1] : undefined;
```
I think we should be setting an error here if undefined comes in

-----
```
const [, outputAmount] = swapRates; 
```

seems much more readable then the new implementation

valeryV99 replied:
done

Georgyskr commented on apps/web/src/widgets/swap-form/ui/SwapForm.tsx:
```diff
@@ -77,42 +78,43 @@ export const SwapForm = ({
+    const formatted =
+      quote !== undefined ? formatUnits(quote, outputTokenInfo.decimals) : '0';
+    setValue('outputTokenAmount', formatted);
   }, 0);
```
let's increase that to 250, right away, perhaps?

Georgyskr commented on apps/web/src/widgets/swap-form/ui/SwapForm.tsx:
```diff
@@ -122,27 +124,13 @@ export const SwapForm = ({
+    const formatted =
+      quote !== undefined ? formatUnits(quote, inputTokenInfo.decimals) : '0';
+    setValue('inputTokenAmount', formatted);
   }, 0);
```
let's increase that to 250, right away, perhaps?

valeryV99 replied:
done

Georgyskr commented:
Also @valeryV99 
I think there is a possibility of a race condition atm with the current implementation (useReadContract initial request coming in after the refetch(or something else) request)

valeryV99 committed:
refactor(SwapForm): improve swap rates handling and update tests

- Added `cancelRefetch` parameter to `readSwapRates` calls to prevent unnecessary refetching.
- Improved debounced callbacks with a 250ms delay for smoother updates of token amounts.
- Simplified logic for formatting and setting token amounts by handling undefined rates directly.
- Enhanced test coverage for `useSwapRates`, including parameter validation and missing argument cases.
- Refactored `useSwapRates` to consolidate logic and eliminate unnecessary return object.

valeryV99 committed:
refactor(useSwapRates): enhance readability and improve test consistency

- Adjusted `useSwapRates` tests to improve parameter readability through multiline formatting.
- Replaced `mock.calls.at(-1)` with `mock.lastCall` for improved clarity.
- Refactored table-driven tests for better structure and indentation.

opcode0x committed:
fix: cancel queries for swap rates

opcode0x committed:
chore: update tests
